### PR TITLE
Update Compliance Adapter 

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
This PR is intended to update the compliance adapter to work with the latest version of FIDO conformance testing tool (tested with v1.6.41 and v1.6.42). This PR will remove some stale code and will implement some additional logic. Following major changes are done.

- Update authentication and registration adapter implementation.
- Adding reset counter API path.
- Add support to store created users and delete them.
- Moving cors implementations to the main file (app.ts).
- Removing stale code blocks.
- Add capability to work with both identity server and Asgardeo
- Git documentation update

**Related Issue**
- https://github.com/wso2/product-is/issues/13495

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -13,108 +13,106 @@
...
 
+6. Start the adapter by executing following command.
 ```sh
-cd /security
-openssl req -nodes -new -x509 -keyout server.key -out server.cert
+npm start
 ```
 
-2. Allow unauthorized TLS
+## Run FIDO Comformance Tests
```
Comment (ThaminduDilshan) [inline README.md]:
```suggestion
## Run FIDO Conformance Tests
```

File: src/app/registration.ts [hunk 2]
```diff
@@ -1,30 +1,22 @@
...
-import verifySignature = require("@simplewebauthn/server/dist/helpers/verifySignature");
-import toHash = require("@simplewebauthn/server/dist/helpers/toHash");
-import decodeCredentialPublicKey = require("@simplewebauthn/server/dist/helpers/decodeCredentialPublicKey");
-import convertCOSEtoPKCS = require("@simplewebauthn/server/dist/helpers/convertCOSEtoPKCS");
 
+const base64url = require("base64url");
+const userOps = require("./user");
 const getCertificateInfo = require("@simplewebauthn/server/dist/helpers/getCertificateInfo");
 
 let config = require("./../../config.json");
```
Comment (ThaminduDilshan) [inline src/app/registration.ts]: Change config to `const`

File: src/app/registration.ts [hunk 3]
```diff
@@ -1,30 +1,22 @@
...
 
 var challenge: any;
```
Comment (ThaminduDilshan) [inline src/app/registration.ts]: Change var to `let`. Check other places as well

File: src/app/user.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    return token;
+}
+
+const formatUsername = (providedName) => {
+    const validRegex = /^[^!@#$%\^&\*\(\)_\+\-=\[\]{};':"\\|,.<>\/?]+(.)*[^!@#$%\^&\*\(\)_\+\-=\[\]{};':"\\|,.<>\/?]+@(.)+\.(.)+$/;
+    const invalidRegex1 = /[!@#$%\^&\*\(\)_\+\-=\[\]{};':"\\|,.<>\/?]{2,}/;
+
+    if (config.isCloudSetup) {
+        if (providedName.includes("@")) {
+            providedName = `CUSTOMER-DEFAULT/fido${providedName}`;
```
Comment (ThaminduDilshan) [inline src/app/user.ts]: move `CUSTOMER-DEFAULT` to config.json. Add a variable called userStoredomain

File: src/app.ts [hunk 5]
```diff
@@ -1,12 +1,17 @@
 const express = require("express");
```
Comment (ThaminduDilshan) [inline src/app.ts]: add eslint for code formatting

Comment (ThaminduDilshan) [reply src/app.ts]: Added and resolved the reported issues except for console logs

File: src/app/user.ts (hunk 4 above)
Comment (ThaminduDilshan) [reply src/app/user.ts]: Fixed

File: src/app/registration.ts (hunk 3 above)
Comment (ThaminduDilshan) [reply src/app/registration.ts]: Fixed

File: src/app/registration.ts (hunk 2 above)
Comment (ThaminduDilshan) [reply src/app/registration.ts]: Fixed

File: README.md (hunk 1 above)
Comment (ThaminduDilshan) [reply README.md]: Fixed

File: .eslintignore [hunk 6]
```diff
@@ -0,0 +1,3 @@
+node_modules/
+dist/
+data/
```
Comment (thivi) [inline .eslintignore]: Add a new empty line

File: .prettierrc.json [hunk 7]
```diff
@@ -1 +0,0 @@
-{}
```
Comment (thivi) [inline .prettierrc.json:1]: Do we need this file?

File: README.md [hunk 8]
```diff
@@ -13,108 +13,107 @@
 
 ## Setup Guide
 
-### Downloading and Installation - FIDO Conformance tools
+### Setup FIDO Conformance Testing Tool
```
Comment (thivi) [inline README.md]:
```suggestion
### Set up FIDO Conformance Testing Tool
```

File: README.md [hunk 9]
```diff
@@ -13,108 +13,107 @@
...
+3. Launch the application and select **FIDO2 Server - MDS3 Tests**.
 
-Note : Tools (v1.5.2) will work only on macOS and Windows.
+3. Enter `https://localhost:4000` as the server url.
 
-### Adater Setup
+> Note: Tool (v1.6.42 experimental) will work only on macOS and Windows operating systems.
 
-fido2-compliance-adapter requires **Node.js** to run.
+### Setup Identity Server
```
Comment (thivi) [inline README.md]:
```suggestion
### Set up Identity Server
```

File: src/app/authentication.ts [hunk 10]
```diff
@@ -1,139 +1,268 @@
...
+            + `&client_id=${client_id}`;
+
+        // start-authentication.
+        await axios({
+            headers: {
+                Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*",  
+                Authorization: `Basic ${auth}`,
+                Connection: "keep-alive",
+                ContentType: "application/json",
+                Host: "localhost:9443"
```
Comment (thivi) [inline src/app/authentication.ts]: Can we extract this into a constant?

File: src/app/authentication.ts [hunk 11]
```diff
@@ -1,139 +1,268 @@
...
+
+            /**
+             * This particular test expects the allowCredentials list and that is not supported by the browser.
+             * Allow credentials are not required in the usernameless authentication.
+             * According to spec, credentials are used when the user has to be identified.
+             */
+            if (optionsRequestCounter == 1 && resultRequestCounter == 0) {
+                responseToTool.allowCredentials = [
+                    {
+                        id: "rnInB99skrSHLwQJpAio3W2S5RMHGYGudqdobiUImDI",
```
Comment (thivi) [inline src/app/authentication.ts]: Can we extract this into a constant?

File: src/app/authentication.ts [hunk 12]
```diff
@@ -1,139 +1,268 @@
...
+
+                /**
+                 * This particular test expects the allowCredentials list and that is not supported by the browser.
+                 * Allow credentials are not required in the usernameless authentication.
+                 * According to spec, credentials are used when the user has to be identified.
+                 */
+                if (optionsRequestCounter == 1 && resultRequestCounter == 0) {
+                    responseToTool.allowCredentials = [
+                        {
+                            id: "rnInB99skrSHLwQJpAio3W2S5RMHGYGudqdobiUImDI",
```
Comment (thivi) [inline src/app/authentication.ts]: Can we make this a constant?

File: src/app/authentication.ts [hunk 13]
```diff
@@ -1,139 +1,268 @@
...
+            delete req.body["authenticatorAttachment"];
         }
-      })
-      .catch((error) => {
+
+        const tr = JSON.stringify({ credential: req.body, requestId: requestId });
+        const referer = `https://${config.authRequestRefererHost}`
+            + (config.tenantName && config.tenantName !== "" ? `/t/${config.tenantName}/` : "/") 
+            + "authenticationendpoint/fido2-auth.jsp?authenticators=FIDOAuthenticator%3ALOCAL"
+            + `&type=fido&sessionDataKey=${sessionDataKey}&data=${tr}`;
```
Comment (thivi) [inline src/app/authentication.ts]: Shall we use one of templates literals or double quotes only?

File: src/app/registration.ts [hunk 14]
```diff
@@ -1,485 +1,274 @@
...
+const config = require("./../../config.json");
+const userOps = require("./user");
 
-let config = require("./../../config.json");
-
-var challenge: any;
-var auth: any;
+let authClient: any;
+let token: any;
+let invalidUsername: any = false;
```
Comment (thivi) [inline src/app/registration.ts]: Shouldn't this be boolean?

File: .prettierrc.json (hunk 7 above)
Comment (ThaminduDilshan) [reply .prettierrc.json:1]: No. I have removed it

File: .eslintignore (hunk 6 above)
Comment (ThaminduDilshan) [reply .eslintignore]: Fixed

File: README.md (hunk 8 above)
Comment (ThaminduDilshan) [reply README.md]: Fixed

File: README.md (hunk 9 above)
Comment (ThaminduDilshan) [reply README.md]: Fixed

File: src/app/authentication.ts (hunk 10 above)
Comment (ThaminduDilshan) [reply src/app/authentication.ts]: Fixed

File: src/app/authentication.ts (hunk 11 above)
Comment (ThaminduDilshan) [reply src/app/authentication.ts]: Fixed

File: src/app/authentication.ts (hunk 12 above)
Comment (ThaminduDilshan) [reply src/app/authentication.ts]: Fixed

File: src/app/authentication.ts (hunk 13 above)
Comment (ThaminduDilshan) [reply src/app/authentication.ts]: Fixed

File: src/app/registration.ts (hunk 14 above)
Comment (ThaminduDilshan) [reply src/app/registration.ts]: Fixed

**Review by thivi — APPROVED** [review]

**Review by ashensw — APPROVED** [review]

**Pull request merged**